### ISSUE TITLE
Wrong latent after reloading AnnData 

### ISSUE THREAD

**Issue by pchiang5** [YosefLab/scVI#577]
Hi,

**After training, I wrote the raw data plus analyses as below:**

```python
post_adata = anndata.AnnData(all_dataset.X)
post_adata.var['gene_names'] = all_dataset.gene_names
post_adata.obs['CellID'] = all_dataset.CellID
post_adata.obsm["X_scVI"] = latent  #verified correct latent here
post_adata.write_loom('XXX.loom', write_obsm_varm=True)
```

**Then I found the latent reloaded by get_latent() was different from the correct all_dataset.X_scVI:**
```python
all_dataset = LoomDataset("XXX.loom", save_path= save_path)
full = trainer.create_posterior(trainer.model, all_dataset, indices=np.arange(len(all_dataset)))
latent, batch_indices, labels = full.sequential().get_latent()
```
**How shall I export/reload the data to have a correct latent location in the loaded GeneExpressionDataset object? Thanks**

**Comment by adamgayoso** [comment]
Which latent distribution are you using? If you're using `latent_distribution="ln"` then the `get_latent()` function is random.

**Comment by pchiang5** [comment]
**I did not change both of the "latent_distribution" and "model.eval()", so probably by default. Below is the full code I used. Where shall I set the parameters to save it correctly?**



all_dataset = GeneExpressionDataset()
all_dataset.populate_from_datasets([gas, emb, blood, mca, tmss2, tmdroplet ])
n_epochs = 1000 if n_epochs_all is None else n_epochs_all
lr = 1e-3
vae = VAE(all_dataset.nb_genes, n_batch=all_dataset.n_batches, n_labels=all_dataset.n_labels,
          n_hidden=128, n_latent=15, n_layers=2, dropout_rate=0.1, dispersion="gene")
trainer = UnsupervisedTrainer(vae, all_dataset, use_cuda=True, n_epochs_kl_warmup=None)
trainer.train(n_epochs=n_epochs)

post_adata = anndata.AnnData(all_dataset.X)
post_adata.var['gene_names'] = all_dataset.gene_names
post_adata.obs['CellID'] = all_dataset.CellID
post_adata.obsm["X_scVI"] = latent  #verified correct latent here
post_adata.write_loom('XXX.loom', write_obsm_varm=True)
all_dataset = LoomDataset("XXX.loom", save_path= save_path)
full = trainer.create_posterior(trainer.model, all_dataset, indices=np.arange(len(all_dataset)))
latent, batch_indices, labels = full.sequential().get_latent()

**Comment by adamgayoso** [comment]
It's hard for me to follow this code because I can't see where you originally assigned the variable `latent`.

**Comment by adamgayoso** [comment]
Also, according to the first part of your issue you refer to `all_dataset.X_scVI`, where `all_dataset` is an scVI gene expression dataset object. You should not be able to access the latent matrix like this. Can you do `print(all_dataset.X_scVI.shape)`?

**Comment by pchiang5** [comment]
> It's hard for me to follow this code because I can't see where you originally assigned the variable `latent`.
> 
Apologies, I missed the following 2 lines in my original post:
```
full = trainer.create_posterior(trainer.model, all_dataset, indices=np.arange(len(all_dataset)))
latent, batch_indices, labels = full.sequential().get_latent()
```
> You should call
> 
> ```
> latent1, batch_indices, labels = full.sequential().get_latent()
> latent2, batch_indices, labels = full.sequential().get_latent()
> ```
> 
> and check if they are equal.

latent1 and 2 are identical
```
>>> latent1
array([[ 0.01842361, -1.0608128 ,  1.2754519 , ..., -0.00394255,
         0.00279737, -0.01069351],
       [-0.00199458, -1.2725018 ,  0.44536507, ..., -0.00460265,
         0.00825   ,  0.00619126],
       [-0.02177836, -0.15482295, -1.5421815 , ..., -0.01512069,
        -0.00941455,  0.00237193],
       ...,
       [-0.0061117 ,  0.12589303,  0.3376484 , ..., -0.00530816,
        -0.00268744, -0.00869179],
       [-0.00890197,  0.21190253,  1.0821619 , ..., -0.00288734,
         0.02157757,  0.00674587],
       [-0.01220865, -0.43788683, -1.808285  , ..., -0.01724116,
         0.0083754 ,  0.00725697]], dtype=float32)
>>> latent2
array([[ 0.01842361, -1.0608128 ,  1.2754519 , ..., -0.00394255,
         0.00279737, -0.01069351],
       [-0.00199458, -1.2725018 ,  0.44536507, ..., -0.00460265,
         0.00825   ,  0.00619126],
       [-0.02177836, -0.15482295, -1.5421815 , ..., -0.01512069,
        -0.00941455,  0.00237193],
       ...,
       [-0.0061117 ,  0.12589303,  0.3376484 , ..., -0.00530816,
        -0.00268744, -0.00869179],
       [-0.00890197,  0.21190253,  1.0821619 , ..., -0.00288734,
         0.02157757,  0.00674587],
       [-0.01220865, -0.43788683, -1.808285  , ..., -0.01724116,
         0.0083754 ,  0.00725697]], dtype=float32)
```

**Comment by pchiang5** [comment]
> Also, according to the first part of your issue you refer to `all_dataset.X_scVI`, where `all_dataset` is an scVI gene expression dataset object. You should not be able to access the latent matrix like this. Can you do `print(all_dataset.X_scVI.shape)`?
> 

print(all_dataset.X_scVI.shape)
(258274, 15)


> Please use the [anndata function](https://anndata.readthedocs.io/en/latest/anndata.read_loom.html#anndata.read_loom) to reload the loom file. Then load the anndata into a scVI dataset using `AnnDatasetFromAnnData`. The `LoomDataset` is not designed to provide access to everything stored in the loom file.

Thanks, the strategy worked. I found saving and loading the train weights are also required to reproduce the plot:
```
torch.save(trainer.model.state_dict(),save_path+'harmonization.vae.allgenes.30.model.pkl')
trainer.model.load_state_dict(torch.load(save_path+'harmonization.vae.allgenes.30.model.pkl'))
```
Is it possible to save the complete scVI (GeneExpressionDataset) object to anndata/loom so I do not have to convert it to anndata and to add gene names, CellID, and latent information individually?

**Comment by adamgayoso** [comment]
First, it's very critical that you do 

```
torch.save(trainer.model.state_dict(),save_path+'harmonization.vae.allgenes.30.model.pkl')
trainer.model.load_state_dict(torch.load(save_path+'harmonization.vae.allgenes.30.model.pkl'))
trainer.model.eval()
```

Second, no there is currently no automatic way, but we will add save and load methods in the future.

**Comment by pchiang5** [comment]
thank you @adamgayoso
